### ISSUE TITLE
Upgrade from 14.0 to 16.0， The App launch time has increased 100ms

### ISSUE THREAD

**Issue by xidipeng** [facebook/facebook-ios-sdk#2193]
### Checklist before submitting a bug report

- [X] I've updated to the latest released version of the SDK
- [X] I've searched for existing [GitHub issues](https://github.com/facebook/facebook-ios-sdk/issues)
- [X] I've looked for existing answers on [Stack Overflow](https://facebook.stackoverflow.com), the [Facebook Developer Community Forum](https://developers.facebook.com/community/) and the [Facebook Developers Group](https://www.facebook.com/groups/fbdevelopers)
- [X] I've read the [Code of Conduct](https://github.com/facebook/facebook-ios-sdk/blob/main/CODE_OF_CONDUCT.md)
- [X] This issue is not security related and can safely be disclosed publicly on GitHub

### Xcode version

14.2.0

### Facebook iOS SDK version

16.0.1

### Dependency Manager

CocoaPods

### SDK Framework

FBSDKLoginKit

### Goals

Upgrade from SDK 14.0 to 16.0， My project is OC，The App launch time has increased 100ms.

![FB SDK](https://user-images.githubusercontent.com/10096995/230707873-aa7ac7be-ad97-4182-8956-90b2483d6dd4.jpg)

### Expected results

Hope to optimize the SDK in launch.

### Actual results

no stack trace

**Comment by nolimitdev** [comment]
Really someone initializes FB SDK as recommended in doc? I immediately knew that it is not good idea. I do it after user request sign in via facebook only...

**MyLoginManager.swft**
```
import FBSDKLoginKit;
import Foundation;
import GoogleSignIn;

class MyLoginManager: NSObject, LoginButtonDelegate {

...

    public func initAppleLogin() {
        ...
    }

    public func initGoogleLogin() {
        ...
    }

    public func initFacebookLogin() {
        // Initilise FB SDK
        FBSDKCoreKit.ApplicationDelegate.shared.application(UIApplication.shared, didFinishLaunchingWithOptions: nil);

        // Logout before sign in to avoid dialog that user already logged in
        let facebookLoginManager = FBSDKLoginKit.LoginManager();
        facebookLoginManager.logOut();
        
        // Fake login button
        let loginButton = FBLoginButton(frame: .zero);
        loginButton.permissions = ["public_profile", "email"];
        loginButton.delegate = self;
        loginButton.isHidden = true;
        loginButton.sendActions(for: .touchUpInside);
    }

    func loginButton(_ loginButton: FBLoginButton, didCompleteWith result: FBSDKLoginKit.LoginManagerLoginResult?, error: Error?) -> Void {
        ...
    }

    func loginButtonDidLogOut(_ loginButton: FBLoginButton) -> Void {
        return;
    }

    ...

}
```

**Comment by eXhausted** [comment]
Any updates? latest version adds even more, we have increase up to 200ms.

**Comment by Zhiyi-Zhang** [comment]
Thanks for reporting the issue. I recently started working on the SDK and saw this issue. I will look into this and see what improvements we can do for this.

**Comment by Zhiyi-Zhang** [comment]
I think some findings I have is:

The SDK does too much synchronous I/O on the main thread during didFinishLaunchingWithOptions. The biggest offenders:

  1. Domain config load (iOS 14.5+) — can block init behind a network request (up to 4s timeout). Didn't exist in v14.
  2. 2 keychain reads — access token + auth token loaded eagerly at init, not on first use (~20-200ms)
  3. 4 NSKeyedUnarchiver deserializations — domain config, server config, gatekeepers, app events config — all synchronous on main thread

  The fix would be to defer non-essential work: load tokens lazily on first access instead of eagerly, and move cache deserialization off the main thread. No behavior change — just delayed execution until the data is actually needed.

Since the change can be big, I will take a somewhat slow path to rollout the changes.

**Comment by xidipeng** [comment]
> I think some findings I have is:
> 
> The SDK does too much synchronous I/O on the main thread during didFinishLaunchingWithOptions. The biggest offenders:
> 
> 1. Domain config load (iOS 14.5+) — can block init behind a network request (up to 4s timeout). Didn't exist in v14.
> 2. 2 keychain reads — access token + auth token loaded eagerly at init, not on first use (~20-200ms)
> 3. 4 NSKeyedUnarchiver deserializations — domain config, server config, gatekeepers, app events config — all synchronous on main thread
> 
> The fix would be to defer non-essential work: load tokens lazily on first access instead of eagerly, and move cache deserialization off the main thread. No behavior change — just delayed execution until the data is actually needed.
> 
> Since the change can be big, I will take a somewhat slow path to rollout the changes.

Looking forward to your good news. Please let us know as soon as there are any new versions available

**Comment by Zhiyi-Zhang** [comment]
> > I think some findings I have is:
> > The SDK does too much synchronous I/O on the main thread during didFinishLaunchingWithOptions. The biggest offenders:
> > 
> > 1. Domain config load (iOS 14.5+) — can block init behind a network request (up to 4s timeout). Didn't exist in v14.
> > 2. 2 keychain reads — access token + auth token loaded eagerly at init, not on first use (~20-200ms)
> > 3. 4 NSKeyedUnarchiver deserializations — domain config, server config, gatekeepers, app events config — all synchronous on main thread
> > 
> > The fix would be to defer non-essential work: load tokens lazily on first access instead of eagerly, and move cache deserialization off the main thread. No behavior change — just delayed execution until the data is actually needed.
> > Since the change can be big, I will take a somewhat slow path to rollout the changes.
> 
> Looking forward to your good news. Please let us know as soon as there are any new versions available

I have brought this up to the whole SDK team and we are targeting this as high priority for this half. 
We believe that there are some I/O heavy operations can be done in the background thread in a non-blocking manner.

**Comment by swinard-meta** [comment]
I've begun looking into this issue, but might need some more information to best target a fix. From the initial screenshot, it appears the performance regression is mostly in dyld and not necessarily runtime code.


@xidipeng 
The regression seems to be in the pre-main() dyld phase (263ms vs ~176ms), not didFinishLaunching (72ms), so I want to confirm a few things before digging in.
  - Podfile linkage:  use_frameworks! (dynamic), use_frameworks! :linkage => :static, or static libs (no use_frameworks!)? Paste the lines if you can.
  - Does the app already pull in Swift elsewhere, or is FBSDK your first Swift dependency?
  - How many dynamic frameworks does the app embed total?
  - Which FBSDK modules do you link -- just FBSDKCoreKit, or Login/Share/Gaming too?
  - Can you grab DYLD_PRINT_STATISTICS=1 output for a cold launch? (scheme → Run → Arguments). That splits the 263ms into dylib load vs. rebase/bind vs. ObjC/Swift setup vs. initializers. If on iOS 26, this may be deprecated, but you can drill into the dyld phase in the same App Launch trace that you used before.
  - To confirm, this was from a cold start?

If it lands in dylib loading, one thing you can check now: make sure you're only linking the FBSDK modules you actually use — each extra dynamic framework adds load time. Static linking would cut it further, but we'd need to add support for that, so let's try to confirm with the dyld sub-breakdown (if you can provide with your setup, that'd be helpful).

---

@eXhausted can you provide any more info regarding the additional 200ms you mentioned?
  - Integration + linkage: SPM / CocoaPods / Carthage / manual, dynamic or static?
  - Version you upgraded from/to, and which modules you link?
  - App ObjC, Swift, or mixed?
  - Any launch phase breakdown:Xcode Organizer launch metrics, Instruments App Launch template, or DYLD_PRINT_STATISTICS=1 for a cold launch?

Mainly we need to know whether the added time is pre-main() (dynamic linking) or in didFinishLaunching

**Comment by swinard-meta** [comment]
I investigated SDK initialization slowness from two angles: runtime initialization work performed in the code itself, and costs associated with pre-main loading and system interaction.

### What we found in code
Two things dominated in code, both in initializeSDK:
1. **Init blocked on the network.** Loading the domain configuration waited on a Graph round-trip before the SDK reported ready, even though routing only ever needs the cached (or ATT-safe default) configuration. The fetch was doing work nothing was waiting for.
2. **SDK setup ran inside the launch frame.** `doSDKSetup` executed during app launch rather than after the first frame.

There were also a few very small, backwards incompatible changes that weren’t worth it.

### Relevant changes in 18.1.1
- Domain configuration resolves from cache/default immediately; the refresh moved off the init path
- SDK setup deferred past the first frame

Measured on an iPhone 17 on a real network and with a valid app ID.
| Metric | Before (warm / cold) | After (warm / cold) | 
| -- | -- | -- | 
| Time-to-SDK-ready | ~106ms / ~177ms | ~25ms / ~37ms (−76% / −79%) | 
| Synchronous init frame | ~13ms / ~23ms | ~10ms / ~17ms (−23% / −30%) |

These are healthy-network figures on one device, so treat them as indicative. They also understate the change: the domain-configuration request carries a 4-second timeout, and initialization used to wait on it, so on a slow or unreachable network the SDK could take that long to report ready. It now resolves from cache or the ATT-safe default (opt-out) immediately and refreshes in the background.

That delay was not confined to SDK-internal work. On iOS 14.5 and above, a Graph request is queued until the domain configuration resolves. So requests your app makes through the SDK during launch sit in that queue until initialization finished, and anything your UI renders from a Graph response waits behind the same timeout. Nothing was lost, since the queue was flushed, but the latency was real. That window is now bounded by the first frame.

### On the dynamic-linking / dyld point
The initial report screenshot showed apparent slowness attributed to dyld. Building the same app twice, once with the SDK as dynamic frameworks and once statically linked, the static build saves roughly **6ms in the dyld phase and ~8ms on a recurring cold launch**, and **around 50ms on the first launch** after install or upgrade for one-off Mach-O validation against a cold dyld cache (again, iPhone 17).
Note that none of the initialization profiling ever approached the 200+ms dyld present in the initial report, which, as it’s 3+ years old now, was likely done on older and less efficient hardware. Also, much of the cost is not mitigable as it’s related to system-level fix-ups and Swift metadata/registration, which is present regardless of dynamic vs static.

**We are not changing the distribution for this right now.** The recurring savings is single-digit milliseconds on the latest hardware, while adding additional static distribution is a maintenance cost we don’t think is worth the ROI currently. Apps that want it can already link the SDK statically today, via `FacebookSDK-Static_XCFramework.zip` on the release.

### Next steps
Closing this out. The blocking-network and launch-frame issues are fixed in **18.1.1**.
If you are still seeing slowness or regressions, please open a new issue with:
- SDK version and how you integrate (SPM / CocoaPods / Carthage, static or dynamic)
- Device and iOS version
- Whether it is warm, recurring-cold (after a reboot), or first-launch-after-install. These differ a lot and are easy to conflate
- Whether other SDKs/libraries are linked in the app (dyld is an aggregate)
- An Instruments **App Launch** trace if you can share one

Thank you!
cc @Zhiyi-Zhang
